### PR TITLE
feat: tag PreCompact checkpoints

### PR DESCRIPTION
**Pull request by EtanHey**
## Summary
- add `chunk_origin` tagging for PreCompact checkpoints without repurposing the overloaded `source` column
- backfill existing checkpoint-shaped rows during VectorStore schema init and log the migration count in `schema_migrations`
- exclude checkpoint chunks from default search, add `include_checkpoints`, and expose `brain_resume(session_id=None, lookback_days=7)` for explicit recovery

## Verification
- `pytest tests/test_precompact_chunk_origin.py tests/test_search_filter_params.py tests/test_think_recall_integration.py::TestMCPToolCount -q` → 29 passed
- `pytest tests/test_precompact_chunk_origin.py tests/test_search_filter_params.py tests/test_hybrid_search.py tests/test_watcher_bridge.py tests/test_chunk_lifecycle.py tests/test_3tool_aliases.py -q` → 135 passed
- `./scripts/run_tests.sh` → 1860 passed, 9 skipped, 75 deselected, 1 xfailed; MCP registration 3 passed; isolated pytest 32 passed; bun 1 pass; regression shell passed
- pre-push hook reran `./scripts/run_tests.sh` with the same passing counts
- snapshot backfill on `/tmp/brainlayer-fm6-precompact-snapshot.db` → 51 checkpoint rows tagged, 0 manual checkpoint-storage note false positives
- touched-file `ruff check` and `ruff format --check` passed

## Notes
- Whole-repo `ruff check` / `ruff format --check` still report pre-existing script lint/format debt outside this PR; touched files are clean.
- `cr review --plain` could not start locally because the CLI is not connected to the repository's CodeRabbit organization; requesting CodeRabbit on the PR instead.

---

> [!NOTE]
> **Medium Risk**
> Introduces a schema migration/backfill (`chunks.chunk_origin`) and changes default search/KG retrieval behavior to exclude checkpoint content, which could affect existing workflows and query results; adds additional retry/overfetch logic around SQLite reads that may impact performance under load.
> 
> **Overview**
> Adds first-class `chunk_origin` classification (new `chunks.chunk_origin` column + `chunk_origin.py` detection helpers) and tags incoming data across ingestion paths (store, watcher/queue drain, upserts, chunk updates), including a one-time backfill migration recorded in `schema_migrations`.
> 
> Changes default retrieval to **exclude** `precompact_checkpoint` chunks across hybrid/text/vector search, exact chunk-id lookups, and KG fact queries unless `include_checkpoints=true`, with KNN overfetch/caching tweaks to avoid starving results when checkpoints are filtered.
> 
> Exposes an explicit `brain_resume` MCP tool to fetch recent PreCompact checkpoints (optionally filtered by `session_id`, with lock-resilient retries) and wires the new parameter/tool through MCP schemas and tests.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8c758a1dd6ed5517a52983b9ef2f868dea66c332. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Add PreCompact checkpoint tagging, filtering, and `brain_resume` tool to VectorStore search
> - Introduces a `chunk_origin` column to the `chunks` table with a backfill migration that detects and tags existing PreCompact checkpoint content using content-marker heuristics in [`chunk_origin.py`](https://github.com/EtanHey/brainlayer/pull/285/files#diff-08baa51d09137d24a69627a4cb8e994e994260293471bb8c6246165674734083).
> - All write paths ([`store.py`](https://github.com/EtanHey/brainlayer/pull/285/files#diff-3f4ff41bcf437c9376f0c0bc397eca30e1b05c0d75d4e95ef1579366d7c4c707), [`vector_store.py`](https://github.com/EtanHey/brainlayer/pull/285/files#diff-74a1a6036a2c74ec470f2d6ebdef8790a9db1c9b3f914616f295e40e7eb73261), [`drain.py`](https://github.com/EtanHey/brainlayer/pull/285/files#diff-b038e117171bfa3617d2351297309d808c2b5cc52995c7572904030beaf1e12c), [`watcher_bridge.py`](https://github.com/EtanHey/brainlayer/pull/285/files#diff-9c9df603fa7a5ecf5cdb0da9ab4077d520d812083da4b052847100f2874f60c9)) now compute and persist `chunk_origin` on insert or update.
> - Vector, binary, and hybrid KNN searches exclude `precompact_checkpoint` chunks by default, overfetching to compensate and truncating to the requested result count; an `include_checkpoints` flag opts back in.
> - Adds a new `brain_resume` MCP tool in [`search_handler.py`](https://github.com/EtanHey/brainlayer/pull/285/files#diff-1f952ebc3cd8256a9bc05612e63df4c5cdcc59336a777da97375fa88fdb8bac0) that queries recent PreCompact checkpoints filtered by `lookback_days` and optional `session_id`, with retry logic on SQLite busy/locked errors.
> - KG fact queries and exact chunk-id lookups also respect the checkpoint exclusion by default.
> - Risk: existing databases undergo a one-time backfill migration on first open; checkpoint count caching (keyed on `PRAGMA data_version`) adds reads on each KNN query when no cached value exists.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 8c758a1.</sup>
> 
>

## Summary by CodeRabbit

* **New Features**
  * Automatic chunk-origin tagging and persistence, plus a one-time migration/backfill to populate origin data
  * Search controls: new include_checkpoints flag (defaults to exclude) and overfetch logic so exclusions don’t starve results
  * New brain_resume tool to return recent session checkpoint content
  * Queue and ingestion now propagate chunk-origin metadata

* **Tests**
  * Extensive end-to-end coverage for detection, storage, migration, search filtering/overfetch, resume, and edge cases

[Review Change Stack](https://app.coderabbit.ai/change-stack/EtanHey/brainlayer/pull/285?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/brainlayer/mcp/search_handler.py [hunk 1]
```diff
@@ -539,6 +557,7 @@ async def _brain_search(
             sentiment,
             source_filter,
             correction_category,
+            include_checkpoints,
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/mcp/search_handler.py]: Fixed in c0529bba: include_checkpoints is no longer part of has_active_filters, so it widens checkpoint visibility without suppressing entity detection or KG routing.

File: src/brainlayer/vector_store.py [hunk 2]
```diff
@@ -198,10 +213,49 @@ def _init_db(self) -> None:
...
+                    OR LOWER(SUBSTR(content, 1, 1024)) LIKE '%"content": "[precompact checkpoint]%'
+                    OR LOWER(SUBSTR(content, 1, 1024)) LIKE '%''content'': ''[precompact checkpoint]%'
+                  )
+                """,
+                (
+                    CHUNK_ORIGIN_PRECOMPACT_CHECKPOINT,
+                    CHUNK_ORIGIN_UNKNOWN,
+                    CHUNK_ORIGIN_PRECOMPACT_CHECKPOINT,
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/vector_store.py:272]: Fixed in c0529bba: migration backfill now trims ASCII whitespace characters including tabs/newlines before prefix matching, matching the Python lstrip behavior for legacy rows.

File: src/brainlayer/search_repo.py [hunk 3]
```diff
@@ -355,6 +358,8 @@ def search(
             if correction_category:
                 where_clauses.append("c.id IN (SELECT chunk_id FROM chunk_tags WHERE tag LIKE ?)")
                 filter_params.append(f"correction:{correction_category}%")
+            if not include_checkpoints:
+                where_clauses.append("COALESCE(c.chunk_origin, 'unknown') != 'precompact_checkpoint'")
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/search_repo.py]: Fixed in c0529bba: default checkpoint exclusion now triggers KNN over-fetch for both float and binary vector search paths. Added regressions for checkpoint-nearest-neighbor windows.

File: src/brainlayer/mcp/search_handler.py [hunk 4]
```diff
@@ -634,9 +654,86 @@ async def _brain_search(
...
+            header = f"\n### Checkpoint {index}: {item['chunk_id']}"
+            if item.get("date"):
+                header += f" | {item['date']}"
+            if item.get("session_id"):
+                header += f" | session: {item['session_id']}"
+            output_parts.append(header)
+            output_parts.append(item["content"])
+        return ([TextContent(type="text", text="\n".join(output_parts))], structured)
+    except Exception as exc:
+        return _error_result(f"Resume error: {exc}")
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/mcp/search_handler.py]: Fixed in c0529bba: _brain_resume now returns the same unpackable tuple shape on errors as it does on success. Added regression coverage for the exception path.

File: src/brainlayer/search_repo.py [hunk 5]
```diff
@@ -373,6 +378,7 @@ def search(
                 or (source_filter and source_filter != "claude_code")
                 or source_filter_like
                 or correction_category
+                or not include_checkpoints
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/search_repo.py]: Fixed in 97033ec9: default checkpoint filtering no longer makes the old rare-filter 10x KNN overfetch universal. KNN width now stays at n_results when no checkpoints exist, and widens only by the indexed checkpoint count when checkpoint rows may be filtered. Added regression: test_vector_search_does_not_overfetch_for_checkpoint_filter_when_no_checkpoints_exist.

File: src/brainlayer/vector_store.py [hunk 6]
```diff
@@ -1260,7 +1319,14 @@ def upsert_chunks(self, chunks: List[Dict[str, Any]], embeddings: List[List[floa
...
                     conversation_id = COALESCE(excluded.conversation_id, chunks.conversation_id),
                     position = COALESCE(excluded.position, chunks.position),
-                    sender = COALESCE(excluded.sender, chunks.sender)
+                    sender = COALESCE(excluded.sender, chunks.sender),
+                    chunk_origin = CASE
+                        WHEN excluded.chunk_origin IS NOT NULL AND excluded.chunk_origin != 'unknown'
+                            THEN excluded.chunk_origin
+                        WHEN chunks.chunk_origin IS NULL
+                            THEN COALESCE(excluded.chunk_origin, 'unknown')
+                        ELSE chunks.chunk_origin
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/vector_store.py:1352]: Fixed in 97033ec9: upsert_chunks now recomputes chunk_origin when a stable chunk_id receives different content, so a checkpoint row can become unknown/normal content instead of staying checkpoint-tagged. Added regression: test_upsert_chunks_recomputes_origin_when_stable_chunk_content_changes.

File: src/brainlayer/search_repo.py [hunk 7]
```diff
@@ -120,6 +123,32 @@ def _contains_meta_noise(content: Optional[str]) -> bool:
...
+                    (CHUNK_ORIGIN_PRECOMPACT_CHECKPOINT,),
+                )
+                .fetchone()[0]
+                or 0
+            )
+        except (apsw.SQLError, TypeError, IndexError):
+            checkpoint_count = 0
+        if checkpoint_count <= 0:
+            return n_results
+        return min(n_results + checkpoint_count, 1000)
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/search_repo.py]: Fixed in 05d215fc: checkpoint-aware KNN widening no longer has the fixed 1000 cap for checkpoint filtering, so a DB with 1001+ nearer checkpoint vectors still fetches far enough to return normal chunks. Added regression: test_vector_search_does_not_starve_normal_results_after_many_checkpoints.

File: src/brainlayer/mcp/search_handler.py [hunk 8]
```diff
@@ -634,9 +653,87 @@ async def _brain_search(
...
+
+        where_clauses = [
+            "COALESCE(chunk_origin, 'unknown') = ?",
+            "(created_at IS NULL OR created_at >= ?)",
+        ]
+        params: list[Any] = [CHUNK_ORIGIN_PRECOMPACT_CHECKPOINT, cutoff]
+        if session_id:
+            where_clauses.append("(conversation_id = ? OR content LIKE ? OR metadata LIKE ?)")
+            session_like = f"%{session_id}%"
+            params.extend([session_id, session_like, session_like])
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/mcp/search_handler.py:716]: Fixed in 05d215fc: brain_resume now escapes SQL LIKE wildcard characters in session_id and uses ESCAPE for content/metadata matching while keeping exact conversation_id matching. Added regression: test_brain_resume_escapes_session_id_like_wildcards.

File: src/brainlayer/search_repo.py [hunk 9]
```diff
@@ -374,7 +406,7 @@ def search(
                 or source_filter_like
                 or correction_category
             )
-            effective_k = min(n_results * 10, 1000) if needs_overfetch else n_results
+            effective_k = self._effective_knn_k(n_results, bool(needs_overfetch), include_checkpoints)
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/search_repo.py:443]: Fixed in cf6b5f77: direct float and binary vector search now clamp post-overfetch rows back to n_results, so checkpoint-aware k widening preserves the caller result cap. Added regression: test_vector_search_caps_results_after_checkpoint_overfetch.

File: src/brainlayer/search_repo.py [hunk 10]
```diff
@@ -355,6 +385,8 @@ def search(
             if correction_category:
                 where_clauses.append("c.id IN (SELECT chunk_id FROM chunk_tags WHERE tag LIKE ?)")
                 filter_params.append(f"correction:{correction_category}%")
+            if not include_checkpoints:
+                where_clauses.append("COALESCE(c.chunk_origin, 'unknown') != 'precompact_checkpoint'")
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/search_repo.py]: Fixed in cf6b5f77: VectorStore now records whether readonly DBs actually have chunk_origin, and search skips checkpoint-origin predicates when the column is absent, treating legacy rows as unknown. Added regression: test_readonly_legacy_db_without_chunk_origin_searches_as_unknown.

**Comment by EtanHey** [comment]
@coderabbitai review
@codex review
@cursor @bugbot re-review
